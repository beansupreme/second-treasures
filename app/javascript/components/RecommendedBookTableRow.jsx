import React from "react";
import BookTableEditRow from "./BookTableEditRow";
import BookTableViewRow from "./BookTableViewRow";

/* 
  RecommendedBookTableRow manages the state of a single row,
  whether it's editable or not. All other state changes are 
  propogated up to it's parent.
*/
class RecommendedBookTableRow extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      editMode: false
    }
    this.handleDeleteClick = this.handleDeleteClick.bind(this); 
    this.handleEditClick = this.handleEditClick.bind(this); 
    this.handleUpdate = this.handleUpdate.bind(this); 
    this.afterUpdate = this.afterUpdate.bind(this); 
  }

  handleDeleteClick() {
    this.props.onBookDelete(this.props.book.id)
  }

  handleEditClick() {
    this.setState({editMode: true})
  }

  handleUpdate(book) {    
    this.props.onBookUpdate(book, this.afterUpdate);
  }
  // Callback for RecommendedBookTable to call after updating server-side
  // if there are errors, this component will stay in the edit state. 
  afterUpdate(success) {
    if (success) {
      this.setState({editMode: false}) 
    }
  }
  // the state.editMode determines what view should be displayed
  render() {
    return (
      this.state.editMode ?
      <BookTableEditRow book={this.props.book} onUpdate={this.handleUpdate}
      onFieldChange={this.props.onFieldChange}
      /> 
      :
      <BookTableViewRow book={this.props.book} 
      onEditClick={this.handleEditClick} onDeleteClick={this.handleDeleteClick} />
    )
  }  
}

RecommendedBookTableRow.defaultProps = {
  onBookDelete: () => {},
  onBookUpdate: () => {}
}


export default RecommendedBookTableRow;