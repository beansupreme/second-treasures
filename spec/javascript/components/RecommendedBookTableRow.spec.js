import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import RecommendedBookTableRow from 'components/RecommendedBookTableRow';

describe('<RecommendedBookTableRow />', () => {
  let book = {
    id: 6,
    isbn: '978-330020242',
    title: 'Vimy Ridge',
    author: 'Alexander McKee',
    genre: 'War novel',
    price: '1.50',
    createdAt: "2018-06-25T14:44:31.990Z",
    updatedAt: "2018-06-25T14:44:31.990Z"
  }

  it('renders as expected', () => {

    const tree = renderer.create(<RecommendedBookTableRow book={book} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  
  it('Starts out in a non-edit state', () => {
    const wrapper = mount(<RecommendedBookTableRow book={book} />);

    expect(wrapper.state('editMode')).toEqual(false);
  });

  

  describe('when editMode is false', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(<RecommendedBookTableRow book={book} />);
      wrapper.setState({editMode: false})
    });

    it('renders read-only table data', () => {
      let tableData = wrapper.find('td');

      expect(tableData.at(0)).toHaveHTML('<td>Vimy Ridge</td>')
      expect(tableData.at(1)).toHaveHTML('<td>Alexander McKee</td>')
      expect(tableData.at(2)).toHaveHTML('<td><span>$1.50</span></td>')
      expect(tableData.at(3)).toHaveHTML('<td>978-330020242</td>')
      expect(tableData.at(4)).toHaveHTML('<td>War novel</td>')
    });

    it('switches to an editable state when edit is clicked', () => {
      wrapper.find('.edit-book').first().simulate('click');

      expect(wrapper.state('editMode')).toEqual(true);
    });

  });

  describe('when editMode is true', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(<RecommendedBookTableRow book={book} />);
      wrapper.setState({editMode: true})
    });


    it('renders edtiable inputs filled in with the book prop values', () => {
      let tableData = wrapper.find('td');
      
      expect(tableData.at(0).find('input')).toHaveValue('Vimy Ridge')
      expect(tableData.at(1).find('input')).toHaveValue('Alexander McKee')
      expect(tableData.at(2).find('input')).toHaveValue('1.50')
      expect(tableData.at(3).find('input')).toHaveValue('978-330020242')
      expect(tableData.at(4).find('input')).toHaveValue('War novel')
      
    });

    it('renders a save button', () => {
      expect(wrapper.find('.save-book')).toHaveHTML('<button class="btn btn-success save-book">Save</button>')
    });
  })
});