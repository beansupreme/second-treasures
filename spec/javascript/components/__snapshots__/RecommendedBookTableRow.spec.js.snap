// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RecommendedBookTableRow /> renders as expected 1`] = `
<tr
  className="recommended-book-row"
>
  <td>
    Vimy Ridge
  </td>
  <td>
    Alexander McKee
  </td>
  <td>
    <span>
      $
      1.50
    </span>
  </td>
  <td>
    978-330020242
  </td>
  <td>
    War novel
  </td>
  <td>
    <div
      className="btn-group"
      role="group"
    >
      <button
        className="btn btn-warning edit-book"
        onClick={[Function]}
      >
        Edit
      </button>
      <button
        className="btn btn-danger delete-book"
        onClick={[Function]}
      >
        Delete
      </button>
    </div>
  </td>
</tr>
`;
